### PR TITLE
Create a project in datatools

### PR DESCRIPTION
**Pull request by maukoquiroga**
partial_fix #178

### CODE REVIEW

File: lib/transport/data_validation/repository/project_repository.ex [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    end
+  end
+
+  @doc """
+  Creates a project.
+  """
+  @spec create(CreateProject.t) :: {:ok, Project.t} | {:error, any()}
+  def create(%CreateProject{} = command) do
+    case HTTPoison.post(@endpoint, Poison.encode!(command)) do
+      {:ok, %HTTPoison.Response{status_code: 200, body: body}} ->
```
[the remark this answers is not in the document]
Comment (maukoquiroga) [reply lib/transport/data_validation/repository/project_repository.ex]: It _should_ be 201, but it is 200 as you can see in the recorded response (cassette).

File: lib/transport/data_validation/data_validation.ex [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+  alias Transport.DataValidation.Aggregates.Project
+  alias Transport.DataValidation.Commands.CreateProject
+
+  @doc """
+  Creates a new project.
+  """
+  @spec create_project(map()) :: {:ok, Project.t} | {:error, any()}
+  def create_project(%{} = params) do
+    command = CreateProject.new(params)
```
[the remark this answers is not in the document]
Comment (maukoquiroga) [reply lib/transport/data_validation/data_validation.ex]: It is validated, two lines after 😄

File: lib/transport/data_validation/commands/create_project.ex [hunk 3]
```diff
@@ -0,0 +1,31 @@
+defmodule Transport.DataValidation.Commands.CreateProject do
+  @moduledoc """
+  Command for creating a project.
+
+  ## Examples
+
+      iex> %{name: "transport"}
+      ...> |> CreateProject.new
+      ...> |> CreateProject.validate
```
[the remark this answers is not in the document]
Comment (maukoquiroga) [reply lib/transport/data_validation/commands/create_project.ex]: I'll remove `ExConstructor`, as it creates confusion. It should just be a pure function.

[the remark this answers is not in the document]
Comment (maukoquiroga) [reply lib/transport/data_validation/commands/create_project.ex]: Following SRP, I see `new` function as the proper parsing of params (that's what `Exconstructor` does), and `validate` to ensure the _business invariants_ of the contract.

File: lib/transport/data_validation/data_validation.ex (hunk 2 above)
[the remark this answers is not in the document]
Comment (l-vincent-l) [reply lib/transport/data_validation/data_validation.ex]: Sorry I didn't see :confused:

File: lib/transport/data_validation/data_validation.ex [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+    case CreateProject.validate(command) do
+      {:ok, command} -> Project.execute(command)
```
Comment (l-vincent-l) [inline lib/transport/data_validation/data_validation.ex]: don't you wanna do: 
```
params
|> CreateProject.new()
|> CreateProject.validate()
|> case do
{:ok, command} -> …
```
?

File: lib/transport/data_validation/commands/create_project.ex (hunk 3 above)
[the remark this answers is not in the document]
Comment (maukoquiroga) [reply lib/transport/data_validation/commands/create_project.ex]: More precisely I'd like to keep separated the data structure population and validation.

[the remark this answers is not in the document]
Comment (l-vincent-l) [reply lib/transport/data_validation/commands/create_project.ex]: ok

File: lib/transport/data_validation/data_validation.ex (hunk 2 above)
[the remark this answers is not in the document]
Comment (l-vincent-l) [reply lib/transport/data_validation/data_validation.ex]: :confused: 

Just a style thing, in other part of the code we would have written that like it:
```elixir
params
|> CreateProject.new()
|> CreateProject.validate()
|> case do
  {:ok, command} -> Project.execute(command)
…
```

File: lib/transport/data_validation/repository/project_repository.ex [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+      {:error, error} -> {:error, error}
+    end
+  end
+
+  @doc """
+  Creates a project.
+  """
+  @spec create(CreateProject.t) :: {:ok, Project.t} | {:error, any()}
+  def create(%CreateProject{} = command) do
+    case @client.post(@endpoint, Poison.encode!(command)) do
```
Comment (l-vincent-l) [inline lib/transport/data_validation/repository/project_repository.ex]: I think you can use `with` here and don't use a bang function.
```elixir
with {:ok, body} <- Poison.encode(command) 
        {:ok,  %@res{status_code: 200, body: body}} <- @client.post(@endpoint, body) do
  Poison.decode(body, as: %Project{})
else
  {:ok, %@res{body: body}} -> {:error, error}
  {:error, %@err{reason: error}} -> {:error, error}
  {:error, error} -> {:error, error}
end
```

File: lib/transport/data_validation/data_validation.ex (hunk 4 above)
Comment (maukoquiroga) [reply lib/transport/data_validation/data_validation.ex]: OK

File: lib/transport/data_validation/data_validation.ex [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  """
+
+  alias Transport.DataValidation.Aggregates.Project
+  alias Transport.DataValidation.Commands.CreateProject
+
+  @doc """
+  Creates a new project.
+  """
+  @spec create_project(map()) :: {:ok, Project.t} | {:error, any()}
+  def create_project(%{} = params) do
```
Comment (maukoquiroga) [inline lib/transport/data_validation/data_validation.ex]: @l-vincent-l Done 😄

File: lib/transport/data_validation/repository/project_repository.ex (hunk 5 above)
Comment (maukoquiroga) [reply lib/transport/data_validation/repository/project_repository.ex]: Nice one

Comment (maukoquiroga) [reply lib/transport/data_validation/repository/project_repository.ex]: Done

**Pull request merged**